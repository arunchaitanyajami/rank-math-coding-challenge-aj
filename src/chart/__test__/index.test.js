import '@testing-library/jest-dom'
import React from 'react'
import { render, fireEvent, act, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import Chart from '../../chart'
import apiFetch from '@wordpress/api-fetch'

jest.mock('@wordpress/api-fetch')

const data = [
  { day: 'Day 1', value: 10 },
  { day: 'Day 2', value: 10 },
  { day: 'Day 3', value: 60 },
  { day: 'Day 4', value: 30 },
  { day: 'Day 5', value: 70 },
  { day: 'Day 6', value: 70 },
  { day: 'Day 7', value: 100 },
  { day: 'Day 8', value: 25 },
  { day: 'Day 9', value: 39 },
  { day: 'Day 10', value: 69 },
  { day: 'Day 11', value: 69 },
  { day: 'Day 12', value: 69 },
  { day: 'Day 13', value: 69 },
  { day: 'Day 14', value: 69 },
  { day: 'Day 15', value: 69 }
]

describe('Chart Component', function () {
  /**
     * Lets mock hooks.
     */
  let originalUseState

  beforeEach(() => {
    window.graphWidgetSettings = {
      siteUrl: '',
      nonce: 'test-nonce'
    }

    originalUseState = jest.spyOn(React, 'useState')
    apiFetch.mockResolvedValueOnce(Promise.resolve(data))
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  test('Needs a Snapshot', async () => {
    let renderer

    originalUseState.mockReturnValueOnce([data, jest.fn()])

    await act(async () => {
      renderer = render(<Chart chartHeight={470} />)
    })

    expect(renderer.container).toMatchSnapshot()
  })

  test('Should render properly', async () => {
    let renderer

    originalUseState.mockReturnValueOnce([data, jest.fn()])
    await act(async () => {
      renderer = render(<Chart chartHeight={470} />)
    })

    const { container } = renderer

    expect(container.getElementsByClassName('recharts-wrapper').length).toBe(1)
    expect(container.getElementsByClassName('recharts-cartesian-axis-ticks').length).toBe(2)
  })

  test('Should display default filtered data before changing the filter value', async () => {
    let renderer
    originalUseState.mockReturnValueOnce([data, jest.fn()])
    await act(async () => {
      renderer = render(<Chart chartHeight={470} />)
    })

    const { container } = renderer

    const chartSelect = container.querySelector('#filter-select')
    expect(chartSelect.firstChild.textContent).toBe('Last 7 DAYS')

    const xAxisData = container.querySelectorAll('.recharts-xAxis .recharts-cartesian-axis-tick-line')
    expect(xAxisData.length).toBe(7)

    const xAxisDataFirstChild = container.querySelector(
      '.recharts-xAxis .recharts-cartesian-axis-tick:first-child .recharts-cartesian-axis-tick-value tspan'
    )
    expect(xAxisDataFirstChild.innerHTML).toBe('Day 9')
  })

  test('Should display filtered data when a value changes in the select filter', async () => {
    let renderer
    originalUseState.mockReturnValueOnce([data, jest.fn()])
    originalUseState.mockReturnValueOnce([false, jest.fn()])
    originalUseState.mockReturnValueOnce([15, jest.fn()])

    await act(async () => {
      renderer = render(<Chart chartHeight={470} />)
    })

    const { getByRole, container } = renderer
    const selectFilter = getByRole('combobox')

    fireEvent.change(selectFilter, { target: { value: 15 } })
    await userEvent.selectOptions(selectFilter, '15')

    const xAxisData = container.querySelectorAll('.recharts-xAxis .recharts-cartesian-axis-tick-line')
    expect(xAxisData.length).toBe(15)

    const xAxisDataFirstChild = container.querySelector(
      '.recharts-xAxis .recharts-cartesian-axis-tick:first-child .recharts-cartesian-axis-tick-value tspan'
    )
    expect(xAxisDataFirstChild.innerHTML).toBe('Day 1')
  })
})
