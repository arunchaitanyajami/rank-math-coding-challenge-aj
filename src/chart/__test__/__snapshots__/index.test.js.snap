// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart Component Needs a Snapshot 1`] = `
<div>
  <div
    class="dashboard-widget"
  >
    <div
      class="dashboard-widget-container"
    >
      <div
        class="dashboard-select"
      >
        <div
          class="components-base-control css-ij2po-Wrapper-boxSizingReset ej5x27r4"
        >
          <div
            class="components-base-control__field css-10urnh1-StyledField-deprecatedMarginField ej5x27r3"
          >
            <div
              class="components-flex components-select-control em5sgkm7 css-1dhm0tx-View-Flex-base-ItemsColumn-Root e19lxcc00"
              data-wp-c16t="true"
              data-wp-component="Flex"
            >
              <div
                class="components-input-control__container css-1o1a8cj-Container-containerDisabledStyles-containerWidthStyles em5sgkm6"
              >
                <select
                  class="components-select-control__input css-ec4xne-Select-fontSizeStyles-sizeStyles e1mv6sxx2"
                  data-testid="filter-select"
                  id="filter-select"
                >
                  <option
                    value="7"
                  >
                    Last 7 DAYS
                  </option>
                  <option
                    value="15"
                  >
                    Last 15 DAYS
                  </option>
                  <option
                    value="30"
                  >
                    Last 30 DAYS
                  </option>
                </select>
                <span
                  class="components-input-control__suffix css-oobb1x-Suffix em5sgkm0"
                >
                  <div
                    class="components-spacer components-input-control-suffix-wrapper e1mv6sxx0 css-pbf063-View-classes-InputControlSuffixWrapperWithClickThrough e19lxcc00"
                    data-wp-c16t="true"
                    data-wp-component="InputControlSuffixWrapper"
                  >
                    <div
                      class="css-1j80ye7-DownArrowWrapper e1mv6sxx1"
                    >
                      <svg
                        aria-hidden="true"
                        focusable="false"
                        height="18"
                        viewBox="0 0 24 24"
                        width="18"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M17.5 11.6L12 16l-5.5-4.4.9-1.2L12 14l4.5-3.6 1 1.2z"
                        />
                      </svg>
                    </div>
                  </div>
                </span>
                <div
                  aria-hidden="true"
                  class="components-input-control__backdrop css-14s41nl-BackdropUI-backdropFocusedStyles em5sgkm2"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="recharts-responsive-container"
        style="width: 100%; height: 470px; min-width: 0;"
      >
        <div
          class="recharts-wrapper"
          role="region"
          style="position: relative; cursor: default; width: 470px; height: 470px;"
        >
          <svg
            class="recharts-surface"
            height="470"
            viewBox="0 0 470 470"
            width="470"
          >
            <title />
            <desc />
            <defs>
              <clippath
                id="recharts1-clip"
              >
                <rect
                  height="430"
                  width="400"
                  x="65"
                  y="5"
                />
              </clippath>
            </defs>
            <g
              class="recharts-cartesian-grid"
            >
              <g
                class="recharts-cartesian-grid-horizontal"
              >
                <line
                  fill="none"
                  height="430"
                  offset="[object Object]"
                  stroke="#ccc"
                  stroke-dasharray="3,4"
                  width="400"
                  x="65"
                  x1="65"
                  x2="465"
                  y="5"
                  y1="435"
                  y2="435"
                />
                <line
                  fill="none"
                  height="430"
                  offset="[object Object]"
                  stroke="#ccc"
                  stroke-dasharray="3,4"
                  width="400"
                  x="65"
                  x1="65"
                  x2="465"
                  y="5"
                  y1="327.5"
                  y2="327.5"
                />
                <line
                  fill="none"
                  height="430"
                  offset="[object Object]"
                  stroke="#ccc"
                  stroke-dasharray="3,4"
                  width="400"
                  x="65"
                  x1="65"
                  x2="465"
                  y="5"
                  y1="220"
                  y2="220"
                />
                <line
                  fill="none"
                  height="430"
                  offset="[object Object]"
                  stroke="#ccc"
                  stroke-dasharray="3,4"
                  width="400"
                  x="65"
                  x1="65"
                  x2="465"
                  y="5"
                  y1="112.5"
                  y2="112.5"
                />
                <line
                  fill="none"
                  height="430"
                  offset="[object Object]"
                  stroke="#ccc"
                  stroke-dasharray="3,4"
                  width="400"
                  x="65"
                  x1="65"
                  x2="465"
                  y="5"
                  y1="5"
                  y2="5"
                />
              </g>
              <g
                class="recharts-cartesian-grid-vertical"
              >
                <line
                  fill="none"
                  height="430"
                  offset="[object Object]"
                  stroke="#ccc"
                  stroke-dasharray="3,4"
                  width="400"
                  x="65"
                  x1="65"
                  x2="65"
                  y="5"
                  y1="5"
                  y2="435"
                />
                <line
                  fill="none"
                  height="430"
                  offset="[object Object]"
                  stroke="#ccc"
                  stroke-dasharray="3,4"
                  width="400"
                  x="65"
                  x1="131.66666666666669"
                  x2="131.66666666666669"
                  y="5"
                  y1="5"
                  y2="435"
                />
                <line
                  fill="none"
                  height="430"
                  offset="[object Object]"
                  stroke="#ccc"
                  stroke-dasharray="3,4"
                  width="400"
                  x="65"
                  x1="198.33333333333334"
                  x2="198.33333333333334"
                  y="5"
                  y1="5"
                  y2="435"
                />
                <line
                  fill="none"
                  height="430"
                  offset="[object Object]"
                  stroke="#ccc"
                  stroke-dasharray="3,4"
                  width="400"
                  x="65"
                  x1="265"
                  x2="265"
                  y="5"
                  y1="5"
                  y2="435"
                />
                <line
                  fill="none"
                  height="430"
                  offset="[object Object]"
                  stroke="#ccc"
                  stroke-dasharray="3,4"
                  width="400"
                  x="65"
                  x1="331.6666666666667"
                  x2="331.6666666666667"
                  y="5"
                  y1="5"
                  y2="435"
                />
                <line
                  fill="none"
                  height="430"
                  offset="[object Object]"
                  stroke="#ccc"
                  stroke-dasharray="3,4"
                  width="400"
                  x="65"
                  x1="398.33333333333337"
                  x2="398.33333333333337"
                  y="5"
                  y1="5"
                  y2="435"
                />
                <line
                  fill="none"
                  height="430"
                  offset="[object Object]"
                  stroke="#ccc"
                  stroke-dasharray="3,4"
                  width="400"
                  x="65"
                  x1="465"
                  x2="465"
                  y="5"
                  y1="5"
                  y2="435"
                />
              </g>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis recharts-xAxis xAxis"
            >
              <line
                class="recharts-cartesian-axis-line"
                fill="none"
                height="30"
                orientation="bottom"
                stroke="#666"
                width="400"
                x="65"
                x1="65"
                x2="465"
                y="435"
                y1="435"
                y2="435"
              />
              <g
                class="recharts-cartesian-axis-ticks"
              >
                <g
                  class="recharts-layer recharts-cartesian-axis-tick"
                >
                  <line
                    class="recharts-cartesian-axis-tick-line"
                    fill="none"
                    height="30"
                    orientation="bottom"
                    stroke="#666"
                    width="400"
                    x="65"
                    x1="65"
                    x2="65"
                    y="435"
                    y1="441"
                    y2="435"
                  />
                  <text
                    class="recharts-text recharts-cartesian-axis-tick-value"
                    fill="#666"
                    height="30"
                    orientation="bottom"
                    stroke="none"
                    text-anchor="middle"
                    width="400"
                    x="65"
                    y="443"
                  >
                    <tspan
                      dy="0.71em"
                      x="65"
                    >
                      Day 9
                    </tspan>
                  </text>
                </g>
                <g
                  class="recharts-layer recharts-cartesian-axis-tick"
                >
                  <line
                    class="recharts-cartesian-axis-tick-line"
                    fill="none"
                    height="30"
                    orientation="bottom"
                    stroke="#666"
                    width="400"
                    x="65"
                    x1="131.66666666666669"
                    x2="131.66666666666669"
                    y="435"
                    y1="441"
                    y2="435"
                  />
                  <text
                    class="recharts-text recharts-cartesian-axis-tick-value"
                    fill="#666"
                    height="30"
                    orientation="bottom"
                    stroke="none"
                    text-anchor="middle"
                    width="400"
                    x="131.66666666666669"
                    y="443"
                  >
                    <tspan
                      dy="0.71em"
                      x="131.66666666666669"
                    >
                      Day 10
                    </tspan>
                  </text>
                </g>
                <g
                  class="recharts-layer recharts-cartesian-axis-tick"
                >
                  <line
                    class="recharts-cartesian-axis-tick-line"
                    fill="none"
                    height="30"
                    orientation="bottom"
                    stroke="#666"
                    width="400"
                    x="65"
                    x1="198.33333333333334"
                    x2="198.33333333333334"
                    y="435"
                    y1="441"
                    y2="435"
                  />
                  <text
                    class="recharts-text recharts-cartesian-axis-tick-value"
                    fill="#666"
                    height="30"
                    orientation="bottom"
                    stroke="none"
                    text-anchor="middle"
                    width="400"
                    x="198.33333333333334"
                    y="443"
                  >
                    <tspan
                      dy="0.71em"
                      x="198.33333333333334"
                    >
                      Day 11
                    </tspan>
                  </text>
                </g>
                <g
                  class="recharts-layer recharts-cartesian-axis-tick"
                >
                  <line
                    class="recharts-cartesian-axis-tick-line"
                    fill="none"
                    height="30"
                    orientation="bottom"
                    stroke="#666"
                    width="400"
                    x="65"
                    x1="265"
                    x2="265"
                    y="435"
                    y1="441"
                    y2="435"
                  />
                  <text
                    class="recharts-text recharts-cartesian-axis-tick-value"
                    fill="#666"
                    height="30"
                    orientation="bottom"
                    stroke="none"
                    text-anchor="middle"
                    width="400"
                    x="265"
                    y="443"
                  >
                    <tspan
                      dy="0.71em"
                      x="265"
                    >
                      Day 12
                    </tspan>
                  </text>
                </g>
                <g
                  class="recharts-layer recharts-cartesian-axis-tick"
                >
                  <line
                    class="recharts-cartesian-axis-tick-line"
                    fill="none"
                    height="30"
                    orientation="bottom"
                    stroke="#666"
                    width="400"
                    x="65"
                    x1="331.6666666666667"
                    x2="331.6666666666667"
                    y="435"
                    y1="441"
                    y2="435"
                  />
                  <text
                    class="recharts-text recharts-cartesian-axis-tick-value"
                    fill="#666"
                    height="30"
                    orientation="bottom"
                    stroke="none"
                    text-anchor="middle"
                    width="400"
                    x="331.6666666666667"
                    y="443"
                  >
                    <tspan
                      dy="0.71em"
                      x="331.6666666666667"
                    >
                      Day 13
                    </tspan>
                  </text>
                </g>
                <g
                  class="recharts-layer recharts-cartesian-axis-tick"
                >
                  <line
                    class="recharts-cartesian-axis-tick-line"
                    fill="none"
                    height="30"
                    orientation="bottom"
                    stroke="#666"
                    width="400"
                    x="65"
                    x1="398.33333333333337"
                    x2="398.33333333333337"
                    y="435"
                    y1="441"
                    y2="435"
                  />
                  <text
                    class="recharts-text recharts-cartesian-axis-tick-value"
                    fill="#666"
                    height="30"
                    orientation="bottom"
                    stroke="none"
                    text-anchor="middle"
                    width="400"
                    x="398.33333333333337"
                    y="443"
                  >
                    <tspan
                      dy="0.71em"
                      x="398.33333333333337"
                    >
                      Day 14
                    </tspan>
                  </text>
                </g>
                <g
                  class="recharts-layer recharts-cartesian-axis-tick"
                >
                  <line
                    class="recharts-cartesian-axis-tick-line"
                    fill="none"
                    height="30"
                    orientation="bottom"
                    stroke="#666"
                    width="400"
                    x="65"
                    x1="465"
                    x2="465"
                    y="435"
                    y1="441"
                    y2="435"
                  />
                  <text
                    class="recharts-text recharts-cartesian-axis-tick-value"
                    fill="#666"
                    height="30"
                    orientation="bottom"
                    stroke="none"
                    text-anchor="middle"
                    width="400"
                    x="465"
                    y="443"
                  >
                    <tspan
                      dy="0.71em"
                      x="465"
                    >
                      Day 15
                    </tspan>
                  </text>
                </g>
              </g>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis recharts-yAxis yAxis"
            >
              <line
                class="recharts-cartesian-axis-line"
                fill="none"
                height="430"
                orientation="left"
                stroke="#666"
                width="60"
                x="5"
                x1="65"
                x2="65"
                y="5"
                y1="5"
                y2="435"
              />
              <g
                class="recharts-cartesian-axis-ticks"
              >
                <g
                  class="recharts-layer recharts-cartesian-axis-tick"
                >
                  <line
                    class="recharts-cartesian-axis-tick-line"
                    fill="none"
                    height="430"
                    orientation="left"
                    stroke="#666"
                    width="60"
                    x="5"
                    x1="59"
                    x2="65"
                    y="5"
                    y1="435"
                    y2="435"
                  />
                  <text
                    class="recharts-text recharts-cartesian-axis-tick-value"
                    fill="#666"
                    height="430"
                    orientation="left"
                    stroke="none"
                    text-anchor="end"
                    width="60"
                    x="57"
                    y="435"
                  >
                    <tspan
                      dy="0.355em"
                      x="57"
                    >
                      0
                    </tspan>
                  </text>
                </g>
                <g
                  class="recharts-layer recharts-cartesian-axis-tick"
                >
                  <line
                    class="recharts-cartesian-axis-tick-line"
                    fill="none"
                    height="430"
                    orientation="left"
                    stroke="#666"
                    width="60"
                    x="5"
                    x1="59"
                    x2="65"
                    y="5"
                    y1="327.5"
                    y2="327.5"
                  />
                  <text
                    class="recharts-text recharts-cartesian-axis-tick-value"
                    fill="#666"
                    height="430"
                    orientation="left"
                    stroke="none"
                    text-anchor="end"
                    width="60"
                    x="57"
                    y="327.5"
                  >
                    <tspan
                      dy="0.355em"
                      x="57"
                    >
                      20
                    </tspan>
                  </text>
                </g>
                <g
                  class="recharts-layer recharts-cartesian-axis-tick"
                >
                  <line
                    class="recharts-cartesian-axis-tick-line"
                    fill="none"
                    height="430"
                    orientation="left"
                    stroke="#666"
                    width="60"
                    x="5"
                    x1="59"
                    x2="65"
                    y="5"
                    y1="220"
                    y2="220"
                  />
                  <text
                    class="recharts-text recharts-cartesian-axis-tick-value"
                    fill="#666"
                    height="430"
                    orientation="left"
                    stroke="none"
                    text-anchor="end"
                    width="60"
                    x="57"
                    y="220"
                  >
                    <tspan
                      dy="0.355em"
                      x="57"
                    >
                      40
                    </tspan>
                  </text>
                </g>
                <g
                  class="recharts-layer recharts-cartesian-axis-tick"
                >
                  <line
                    class="recharts-cartesian-axis-tick-line"
                    fill="none"
                    height="430"
                    orientation="left"
                    stroke="#666"
                    width="60"
                    x="5"
                    x1="59"
                    x2="65"
                    y="5"
                    y1="112.5"
                    y2="112.5"
                  />
                  <text
                    class="recharts-text recharts-cartesian-axis-tick-value"
                    fill="#666"
                    height="430"
                    orientation="left"
                    stroke="none"
                    text-anchor="end"
                    width="60"
                    x="57"
                    y="112.5"
                  >
                    <tspan
                      dy="0.355em"
                      x="57"
                    >
                      60
                    </tspan>
                  </text>
                </g>
                <g
                  class="recharts-layer recharts-cartesian-axis-tick"
                >
                  <line
                    class="recharts-cartesian-axis-tick-line"
                    fill="none"
                    height="430"
                    orientation="left"
                    stroke="#666"
                    width="60"
                    x="5"
                    x1="59"
                    x2="65"
                    y="5"
                    y1="5"
                    y2="5"
                  />
                  <text
                    class="recharts-text recharts-cartesian-axis-tick-value"
                    fill="#666"
                    height="430"
                    orientation="left"
                    stroke="none"
                    text-anchor="end"
                    width="60"
                    x="57"
                    y="5"
                  >
                    <tspan
                      dy="0.355em"
                      x="57"
                    >
                      80
                    </tspan>
                  </text>
                </g>
              </g>
            </g>
            <g
              class="recharts-layer recharts-line"
            >
              <path
                class="recharts-curve recharts-line-curve"
                d="M65,225.37499999999997C87.22222222222223,144.74999999999997,109.44444444444446,64.12499999999997,131.66666666666669,64.12499999999997C153.8888888888889,64.12499999999997,176.11111111111111,64.12499999999997,198.33333333333334,64.12499999999997C220.55555555555557,64.12499999999997,242.77777777777777,64.12499999999997,265,64.12499999999997C287.22222222222223,64.12499999999997,309.44444444444446,64.12499999999997,331.6666666666667,64.12499999999997C353.8888888888889,64.12499999999997,376.11111111111114,64.12499999999997,398.33333333333337,64.12499999999997C420.5555555555556,64.12499999999997,442.77777777777777,64.12499999999997,465,64.12499999999997"
                fill="none"
                height="430"
                stroke="#8884d8"
                stroke-dasharray="0px 0px"
                stroke-width="1"
                width="400"
              />
            </g>
          </svg>
          <div
            class="recharts-legend-wrapper"
            style="position: absolute; width: 460px; height: auto; left: 5px; bottom: 5px;"
          >
            <ul
              class="recharts-default-legend"
              style="padding: 0px; margin: 0px; text-align: center;"
            >
              <li
                class="recharts-legend-item legend-item-0"
                style="display: inline-block; margin-right: 10px;"
              >
                <svg
                  class="recharts-surface"
                  height="14"
                  style="display: inline-block; vertical-align: middle; margin-right: 4px;"
                  viewBox="0 0 32 32"
                  width="14"
                >
                  <title />
                  <desc />
                  <path
                    class="recharts-legend-icon"
                    d="M0,16h10.666666666666666
            A5.333333333333333,5.333333333333333,0,1,1,21.333333333333332,16
            H32M21.333333333333332,16
            A5.333333333333333,5.333333333333333,0,1,1,10.666666666666666,16"
                    fill="none"
                    stroke="#8884d8"
                    stroke-width="4"
                  />
                </svg>
                <span
                  class="recharts-legend-item-text"
                  style="color: rgb(136, 132, 216);"
                >
                  value
                </span>
              </li>
            </ul>
          </div>
          <div
            class="recharts-tooltip-wrapper"
            role="dialog"
            style="transform: translate(undefinedpx, undefinedpx); pointer-events: none; visibility: hidden; position: absolute; top: 0px; left: 0px;"
            tabindex="-1"
          >
            <div
              class="recharts-default-tooltip"
              style="margin: 0px; padding: 10px; background-color: rgb(255, 255, 255); border: 1px solid #ccc; white-space: nowrap;"
            >
              <p
                class="recharts-tooltip-label"
                style="margin: 0px;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
